helm/resource_repository.go

```diff
@@ -139,6 +135,15 @@ func resourceRepositoryDelete(d *schema.ResourceData, meta interface{}) error {
 	return nil
 }
 
+func resourceRepositoryExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	err := resourceRepositoryCreate(d, meta)
```

Comment:
Terraform documentation https://godoc.org/github.com/hashicorp/terraform/helper/schema#Resource  recommends not to modify the *ResourceData passed to `Exists` so I'm not convinced about this approach.

From the documentation, it should recreate the resource if you look for it and return false if it doesn't exist. Have you tried that?

Reply:
That wouldn't fix the original problem.

The first PR was an attempt to fix a problem related to repositories not being correctly added to the local repository list, which would cause `terraform plan` (or any other read only command) to fail on a different setup than the one that ran `terraform apply`.

A way to reproduce the original issue is to have a repository setup, then run `terraform apply`, and then on another computer try to run `terraform plan` or `terraform refresh` which would raise an error because the repository wouldn't exist on the local repositories.yml.

I understand that this approach is weird since Exists is only supposed to return whether the resource exists, but a helm repository is not a "remote" resource that is created and persisted in the infrastructure. It's a local resource and it needs to somehow be enforced to exist on the client side.

Since only Exists or Read are called when read only commands are run (like `plan` or `refresh`), it needs to be enforced within one of those. The first attempt was using `Read` but calling `Create` from it causes a recursion since `Create` also calls `Read`.

Reply:
Well, I've made the exact test that you've mentioning using a provider where I've added this `Exists` function:
```go
func resourceRepositoryExists(d *schema.ResourceData, meta interface{}) (bool, error) {
	m := meta.(*Meta)

	_, err := getRepository(d, m)
	if err == ErrRepositoryNotFound {
		return false, nil
	} else {
		return true, err
	}

}
```
and whenever it is run in the second computer the plan adds the creation of the local repo that was not there...

Reply:
I'm not sure I understand. 

My understanding is that `helm_repository` is not making changes to the infrastructure/state, instead it's simply adding to the local `repositories.yaml` file. According to your implementation `terraform plan` or `terraform refresh` would always have different results, depending on the computer that it's run and whether it has been ran before.

This problem is a lot more apparent when you have terraform commands being executed inside containers, for example in CI or projects using docker for uniform development environments.

In general, changes to local configurations that are not tracked by the state should be idempotent and behave the same across all computers, and not have varying plans of execution depending on the local state not tracked by terraform.

On the other hand, if the plan is to deprecate `helm_repository`, then that's a separate discussion which I don't have a lot of context. The fact that my previous PR got merged by another contributor led me to think that this is a problem worth fixing, since it is pretty disruptive and annoying if there are a lot of devs working on the same project or if you're using terraform within containers or any other immutable infrastructure. 

This specific PR is only fixing infinite recursion problem introduced by the [previous change](https://github.com/terraform-providers/terraform-provider-helm/pull/128). If the overall problem is not something that needs to be fixed, then feel free to revert it and we can maintain our own fork.

Reply:
Of course the problem needs to be addressed, we just have different ideas about how to fix it :)

The problem with your approach is that it is doing things that are not recommended by the `terraform` documentation. Furthermore `terraform plan` will be making changes and that must not happen.

I understand that it does not look right that runs in different servers have different plans but it is the way `helm_repository` works as it manages a local resource. At the bottom of the problem is that the resource is flawed for that very same reason.

Following the spirit of your last paragraph, I will close this PR, revert the previous PR and then create a new one with the solution I've proposed as the issue still needs solution. Then I will take a further look to the `helm_repository` resource as it should change to avoid the issues we're experiencing.

I would really like to thank you for your contribution and the time you've spent discussing this issue.